Color decorators for dynamically change their lightness. #7

lemonboston opened the pull request:
@dmfs Please note that `LightenedTest` is failing currently but I don't understand why. If you could check and have an idea, please let me know. Otherwise I will get back to this later.

lemonboston commented:
I've fixed the test by using an actual `android.Color` there as well. Not sure why that to and from HSV version didn't work.
Ready for review.

dmfs commented:
I see one bug: `Toned` doesn't preserve the alpha component of the color.

The other issue is the definition of "lightness". We're using the HSV model (see [HSL and HSV](https://en.wikipedia.org/wiki/HSL_and_HSV#Swatches) for a comparison against HSL. This means by changing the "v" (value) component there is a limit of how "light" the color can become (unless the saturation is 0 you will never get white as a result). Btw, that's why the OT FAB color also depended on the saturation.

In order to make it any lighter when `v==1` you need to increase the saturation. I think we should convert to HSL before applying the function and convert it back afterwards.

I guess that means we're not doing it right in OpenTasks either.

dmfs commented:
Also, I'd like to see a couple of predefined (parametrized) curve functions I can pass to the `Toned` decorator. Often a simple `Math.min` or `Math.max` is not desirable, instead you usually want a softer curve. Sometimes you only want to darken/lighten very light/dark ranges.
So instead of using these `min`/`max` parameters provide some predefined functions.
Check our the "adjust color curve" function in Gimp to see what I mean.

dmfs commented:
Regarding the failing tests, you have to keep in mind that float HSV and int RGB values have limited precision, so the conversion from and to HSV is not lossless. I.e. there will be rounding errors. In particular `HSV -> RGB -> HSV` conversion is unlikely to result in exactly the original HSV values. So when comparing these you have to provide a tolerance other than `0` (without doing the math I'd expect it should be somewhere in the range of `1/256` or `1/512`).

dmfs commented:
How about this idea:

We introduce a new interface:

```java
public interface HslColor
{
   float alpha();
   float hue();
   float saturation();
   float lightness();
}
```
or just 
```java
public interface HslColor
{
   float[] ahsl();
}
```

And with the help of adapters like `RgbHslColor` and `HslRgbColor()` you could write it like this:

```java
Color lighter = new HslRgbColor(new Toned(new LighteningFunction(), new RgbHslColor(color)));
```
i.e. `RgbHslColor` adapts the original color to `HslColor`, which is then `Toned` using the given `Function` and the result is adapted to `Color` using `HslRgbColor`.

`Toned` can only operate on `HSL` so the primary ctor would take the function and a `HslColor`. There could be a secondary ctor which also takes a `Color` instead.

Afaik, Android doesn't provide `HSL` methods but you can see the math on http://www.rapidtables.com/convert/color/rgb-to-hsl.htm so it shouldn't be difficult to write the two adapters for the conversion

lemonboston commented:
I've found this:
https://developer.android.com/reference/android/support/v4/graphics/ColorUtils.html
It has HSL conversion methods. So simply multiplying the Lightness in hsl[2] - will that be good? So basically the same as current implementation, only switching from HSV to HSL.
I've seen that people also simply apply a factor to all the RGB values.

With this utils conversion, I suppose we don't need to complicate with new types.

Regarding the failed test, I know about the conversion precision, it is used in the test actually. But that case was 3 <-> 60 for hsv[0], when I created a color from hsv values, and convert back in the end.

lemonboston commented:
Replaced HSV usage with HSL, and renamed classes to `LinearDarkened` and `LinearLightened` to refer to the function applied.
I suggest having separate ticket for the special functions. (Can you help me pointing to somewhere regarding gimp functions? I've only seen pictures of the UI with that adjust color thing)

I've ignored the tests for now, could you please check if implementation looks okay to you?
(Can we ignore saturation in HSL?)

lemonboston commented:
This is from `java.awt.Color` (can't use this class on Android):
```java
private static final double FACTOR = 0.7;

/**
     * Creates a new <code>Color</code> that is a brighter version of this
     * <code>Color</code>.
     * <p>
     * This method applies an arbitrary scale factor to each of the three RGB
     * components of this <code>Color</code> to create a brighter version
     * of this <code>Color</code>.
     * The {@code alpha} value is preserved.
     * Although <code>brighter</code> and
     * <code>darker</code> are inverse operations, the results of a
     * series of invocations of these two methods might be inconsistent
     * because of rounding errors.
     * @return     a new <code>Color</code> object that is
     *                 a brighter version of this <code>Color</code>
     *                 with the same {@code alpha} value.
     * @see        java.awt.Color#darker
     * @since      JDK1.0
     */
    public Color brighter() {
        int r = getRed();
        int g = getGreen();
        int b = getBlue();
        int alpha = getAlpha();

        /* From 2D group:
         * 1. black.brighter() should return grey
         * 2. applying brighter to blue will always return blue, brighter
         * 3. non pure color (non zero rgb) will eventually return white
         */
        int i = (int)(1.0/(1.0-FACTOR));
        if ( r == 0 && g == 0 && b == 0) {
            return new Color(i, i, i, alpha);
        }
        if ( r > 0 && r < i ) r = i;
        if ( g > 0 && g < i ) g = i;
        if ( b > 0 && b < i ) b = i;

        return new Color(Math.min((int)(r/FACTOR), 255),
                         Math.min((int)(g/FACTOR), 255),
                         Math.min((int)(b/FACTOR), 255),
                         alpha);
    }

    /**
     * Creates a new <code>Color</code> that is a darker version of this
     * <code>Color</code>.
     * <p>
     * This method applies an arbitrary scale factor to each of the three RGB
     * components of this <code>Color</code> to create a darker version of
     * this <code>Color</code>.
     * The {@code alpha} value is preserved.
     * Although <code>brighter</code> and
     * <code>darker</code> are inverse operations, the results of a series
     * of invocations of these two methods might be inconsistent because
     * of rounding errors.
     * @return  a new <code>Color</code> object that is
     *                    a darker version of this <code>Color</code>
     *                    with the same {@code alpha} value.
     * @see        java.awt.Color#brighter
     * @since      JDK1.0
     */
    public Color darker() {
        return new Color(Math.max((int)(getRed()  *FACTOR), 0),
                         Math.max((int)(getGreen()*FACTOR), 0),
                         Math.max((int)(getBlue() *FACTOR), 0),
                         getAlpha());
    }
```

dmfs commented on color-bolts/src/main/java/org/dmfs/android/bolts/color/dynamic/lightness/Toned.java:
```diff
@@ -0,0 +1,62 @@
+
+        hsl[2] = Math.min(Math.max(mLightnessAdjustingFunction.value(hsl[2]), 0), 1);
+
+        return ColorUtils.HSLToColor(hsl);
```
this still doesn't set the alpha of the result. The alpha value should remain the same.

lemonboston replied:
I've updated it to keep the alpha.

lemonboston commented:
Should multiplying the rgb components with the same factor result in the same change as multiplying HSL-Lightness?
That `awt.Color` uses that, and I suppose it should be much faster in our case as well, since the color is coded in rgb originally.
I've pushed a version to `issues/7-dynamic-colors-rgb-multiply` branch. I would expect the same test to pass which check HSL values, but it doesn't. Hue is 0 <-> 180.
(That version would need mix/max application)

lemonboston commented:
Nevermind the rgb-multiplying version, that may not be obvious when reaching the 0 or 255 bounds.
This answer mentions it in the end to just use HSL: https://graphicdesign.stackexchange.com/a/75419

So it's ready for review.

dmfs commented on color-bolts/src/main/java/org/dmfs/android/bolts/color/dynamic/lightness/LinearDarkened.java:
```diff
@@ -0,0 +1,85 @@
+     * @param original
+     *         the color to change
+     */
+    public LinearDarkened(float factor, float minLevel, Color original)
```
please remove the `minLevel` and `maxLevel` arguments. I don't see where we would use that and it looks like a violation of SRP.

dmfs commented on color-bolts/src/main/java/org/dmfs/android/bolts/color/dynamic/lightness/LinearDarkened.java:
```diff
@@ -0,0 +1,85 @@
+    {
+        if (mFactor > 1)
+        {
+            throw new IllegalArgumentException("Darkening multiplying factor has to be >0 and <=1, but was: " + mFactor);
```
The errors message says `>0` but the code doesn't check for that condition.

lemonboston commented:
Ok, moving this back to Backlog while focusing on OT bugs.

lemonboston commented:
It turns out that bounding the value to [0, 1] explicitly is not needed, the conversion does that.
But by the time I noticed this, I've already created a `Clamped` `Function` for that.
I saw the name used here, but also elsewhere as well:
https://developer.android.com/reference/android/support/v4/math/MathUtils.html
Edit: another place where they mention it:
http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2015/n4536.html

Please check if you'd prefer keeping it in `Toned` or just remove it. If we kept it, I would create a ticket in jems, to add `Clamped` and `IdentityFunction` there.

Usage in OT could be:
```
new DarkenedForStatusBar(color); // specific class to re-use it without redefining the factor

class DarkenedForStatusBar implements Color
   ...delegate = new Toned(v -> v * 0.7, color)
```

dmfs commented:
Looking at the sources of `HSLtoColor` I can see that the output RGB values are constrained, but not the input HSL values. Are you sure the result is the same? They undergo some non-linear calculations, I don't think that results are the same.

That being said, you can apply the "Clamped" function right in the ctor, so you don't need to create another instance on every call to `argb`.

We should keep `Clamped` and `IdentityFunction` in any case. We might want to clamp at other values too.

lemonboston commented:
I've submitted `Clamping` and `IdentityFunction` in jems: https://github.com/dmfs/jems/pull/111
but also kept a copy here, marked them deprecated in favor of coming jems version, and reduced their visibility to package-private. So that we don't need a release from jems just because of this in case jitpack is not preferred.
Ready for check.
